### PR TITLE
[go_router] [test-exempt] land change to example (#1383)

### PR DESCRIPTION
**Pull request by johnpryan**
Merge main into https://github.com/flutter/packages/pull/1383 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by chunhtai — COMMENTED** [review]
LGTM

**Comment by chunhtai** [comment]
@johnpryan you may need to rebase to fix the ci

File: packages/go_router/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +10,7 @@
 
 - Refactors runtime checks to assertions.
 
+
```
Comment (johnpryan) [inline packages/go_router/CHANGELOG.md:13]:
```suggestion
```

File: packages/go_router/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+## 3.1.7
```
Comment (chunhtai) [inline packages/go_router/CHANGELOG.md:1]: 0 to 7 seems to be a big jump

File: packages/go_router/example/lib/books/main.dart [hunk 3]
```diff
@@ -62,9 +62,10 @@ class Bookstore extends StatelessWidget {
         redirect: (_) => '/books/popular',
       ),
       GoRoute(
-        path: '/book/:bookId',
-        redirect: (GoRouterState state) =>
-            '/books/all/${state.params['bookId']}',
+        path: '/book/:kind/:bookId',
```
Comment (chunhtai) [inline packages/go_router/example/lib/books/main.dart:65]: are we no longer supporting `/book/:bookId`?

File: packages/go_router/CHANGELOG.md (hunk 2 above)
Comment (chunhtai) [reply packages/go_router/CHANGELOG.md:1]: looks like it should be 3.1.1 based on pubspec.yaml

**Review by chunhtai — CHANGES_REQUESTED** [review]

File: packages/go_router/CHANGELOG.md (hunk 2 above)
Comment (johnpryan) [inline packages/go_router/CHANGELOG.md:1]:
```suggestion
## 3.1.1
```

File: packages/go_router/example/lib/books/main.dart (hunk 3 above)
Comment (johnpryan) [reply packages/go_router/example/lib/books/main.dart:65]: That's my understanding, yes. The state needs to be in the URL in order to be preserved.

Comment (chunhtai) [reply packages/go_router/example/lib/books/main.dart:65]: no a hard requirement inorder to fix this, why not keep both redirects?

**Review by chunhtai — COMMENTED** [review]
Sorry I forgot to mention this before, can you add a test in example/books ? no need to be comprehensive, just a test to verify your change here. We should start adding test for example, too. It should be picked up by the ci automatically.

**Pull request closed without merging**